bokeh/charts/_builder.py

```diff
@@ -0,0 +1,321 @@
+#
+# The full license is in the file LICENCE.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+
+#-----------------------------------------------------------------------------
+# Imports
+#-----------------------------------------------------------------------------
+
+from collections import OrderedDict
+import itertools
+from ._charts import Chart
+from ._data_adapter import DataAdapter
+from ..models import GlyphRenderer
+from ..models.glyphs import (Asterisk, Circle, CircleCross, CircleX, Cross, Diamond,
+                             DiamondCross, InvertedTriangle,
+                             Square, SquareCross, SquareX, Triangle, X)
+
+DEFAULT_PALETTE = ["#f22c40", "#5ab738", "#407ee7", "#df5320", "#00ad9c", "#c33ff3"]
+
+#-----------------------------------------------------------------------------
+# Classes and functions
+#-----------------------------------------------------------------------------
+
+def create_and_build(builder_class, values, **kws):
+    # create the new builder
+    builder = builder_class(values, **kws)
+
+    # create a chart to return, since there isn't one already
+    chart = Chart(**kws)
+    chart.add_builder(builder)
+
+    return chart
+
+
+class Builder(object):
+    """ A prototype class to inherit each new chart Builder type.
+
+    It provides useful methods to be used by the inherited builder classes,
+    in order to automate most of the charts creation tasks and leave the
+    core customization to specialized builder classes. In that pattern
+    inherited builders just need to provide:
+
+     - the following methods:
+        * draw: yields the glyphs to be rendered into the plot (and
+            eventually create the self._legends attribute to be used to
+            create the proper legends when builder is called to build
+            the glyphs on a Chart object
+        * get_data(optional): Get the input data and calculates the 'data'
+            attribute to be used to calculate the source data
+        * get_source(optional): Push data into the self.source attribute
+            (of type ColumnDataSource) and build the proper ranges
+            (self.x_range and self.y_range).
+
+    - the following attributes:
+        x_range:
+        y_range:
+        _legends:
+
+
+    so Builder can use it all to draw on a chart when called with the
+    create method.
+
+    """
+    def __init__(self, values=None, legend=False, palette=None, **kws):
+        """Common arguments to be used by all the inherited classes.
+
+        Args:
+            values (iterable): iterable 2d representing the data series
+                values matrix.
+            legend (str, bool): the legend of your plot. The legend content is
+                inferred from incoming input.It can be ``top_left``,
+                ``top_right``, ``bottom_left``, ``bottom_right``.
+                It is ``top_right`` is you set it as True.
```

Comment:
Should mention that it is optional and default is `False` (which means no legend). Also might possibly be better to link to `Legend` reference or user guide pages for valid locations, instead of duplicating here.